### PR TITLE
Speedup SharedBlobCacheService further

### PR DESCRIPTION
**Pull request by original-brownbear**
* Avoid creating capturing lambda for hot-path region looking
* Made assertion code not run if asserts are off
   * and used proper atomic array instead of manually messing with an array of atomics in assertions
* Extracted cold path for init cache chunk

### CODE REVIEW

**Review by romseygeek — APPROVED** [review]
LGTM

**Review by jpountz — APPROVED** [review]

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java [hunk 1]
```diff
@@ -916,7 +919,7 @@ public boolean tryRead(ByteBuffer buf, long offset) throws IOException {
                 return false;
             }
             final CacheFileRegion fileRegion = get(cacheKey, length, startRegion);
-            if (fileRegion.tracker.checkAvailable(end) == false) {
+            if (fileRegion.tracker.checkAvailable(getRegionRelativePosition(end)) == false) {
```
Comment (tlrx) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:922]: 👍🏻 and the corresponding assertion

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java [hunk 2]
```diff
@@ -446,6 +419,37 @@ public CacheFileRegion get(KeyType cacheKey, long fileLength, int region) {
         return entry.chunk;
     }
 
+    private CacheFileRegion initChunk(Entry<CacheFileRegion> entry) {
+        RegionKey<KeyType> regionKey = entry.chunk.regionKey;
```
Comment (tlrx) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:423]: should we assert assert Thread.holdsLock(entry.chunk)?

**Review by tlrx — APPROVED** [review]
LGTM, left a minor comment

**Pull request merged**